### ISSUE TITLE
Thread safety

### ISSUE THREAD

**Issue by h6ah4i** [bluelinelabs/LoganSquare#50]
Hi. I'm considering to use this library in our product. I have read some part of the implementation and I noticed that this library is not thread-safe for now.

---

First, the implementation of the `LoganSquare.TYPE_CONVERTERS` field is like the following;
```java
private static final Map<Class, TypeConverter> TYPE_CONVERTERS = new HashMap<Class, TypeConverter>();
```

Should be;
```java
private static final Map<Class, TypeConverter> TYPE_CONVERTERS = new ConcurrentHashMap<Class, TypeConverter>();
```

---

Second, the implementation of the `LoganSquare.mapperFor()` method is like the following;

```java
public static <E> JsonMapper<E> mapperFor(Class<E> cls) throws NoSuchMapperException {
    JsonMapper<E> mapper = OBJECT_MAPPERS.get(cls);

    if (mapper == null) {
        try {
            Class<?> mapperClass = Class.forName(cls.getName() + Constants.MAPPER_CLASS_SUFFIX);
            mapper = (JsonMapper<E>)mapperClass.newInstance();
            OBJECT_MAPPERS.put(cls, mapper);
        } catch (Exception e) {
            throw new NoSuchMapperException(cls, e);
        }
    }

    return mapper;
}
```

Should be;
```java
public static <E> JsonMapper<E> mapperFor(Class<E> cls) throws NoSuchMapperException {
    synchronized (OBJECT_MAPPERS) {
        JsonMapper<E> mapper = OBJECT_MAPPERS.get(cls);

        if (mapper == null) {
            try {
                Class<?> mapperClass = Class.forName(cls.getName() + Constants.MAPPER_CLASS_SUFFIX);
                mapper = (JsonMapper<E>)mapperClass.newInstance();
                OBJECT_MAPPERS.put(cls, mapper);
            } catch (Exception e) {
                throw new NoSuchMapperException(cls, e);
            }
        }

        return mapper;
    }
}
```

---

Is that all to make this library thread-safe?

Thanks,
Haruki Hasegawa

**Comment by EricKuck** [comment]
From what I can tell, there's really no reason to make `LoganSquare.TYPE_CONVERTER` a ConcurrentHashMap, as all put operations will take place during app initialization. By the time multiple threads are accessing it, there should never be any writes. Accessing a standard HashMap from multiple threads is perfectly safe, as long as one of those threads isn't also writing to it. Am I missing something?

Regarding the `LoganSquare.mapperFor()` operation, you're totally right. That should be synchronized. I'll do that for the next version.

The rest of the library should be thread safe. We've used it in lots of apps here at BlueLine Labs already and haven't run into a single issue yet.

Thanks!

**Comment by h6ah4i** [comment]
Thank you for your quick response!

> From what I can tell, there's really no reason to make LoganSquare.TYPE_CONVERTER a ConcurrentHashMap, as all put operations will take place during app initialization. By the time multiple threads are accessing it, there should never be any writes. Accessing a standard HashMap from multiple threads is perfectly safe, as long as one of those threads isn't also writing to it. Am I missing something?

Ok, you are right. No need to use `ConcurrentHashMap` for `LoganSquare.TYPE_CONVERTER`.

**Comment by h6ah4i** [comment]
This is an optional suggestion.

Currently, `CalendarTypeConverter` and `CalendarTypeConverter` are using DateFormat with synchronization. I think it can be implemented more better ways.

1. Use `ThreadLocal` + `SimpleDateFormat` (no need to use another libraries)
2. Use Joda-Time's ([joda-time-android](https://github.com/dlew/joda-time-android)) `DateTimeFormatter` class
3. Use Apache Commons' `FastDateFormat` class

The second and the third options depend on another (large footprint...) libraries, so it is not proper to use as the default implementation. However, I think it is a good idea that providing them as separated JARs.

**Comment by EricKuck** [comment]
Move of these changes (as well as a few others to ensure thread safety) are now available on JCenter as version 1.1.0.

**Comment by kwiky** [comment]
I've got an issue with wrong date parsing results which appears randomly using v1.3.4
Examples : 
```
DefaultDateFormatter: 2016-01-27T16:30:00.000Z=>Wed Jan 27 17:30:00 GMT+01:00 2016
DefaultDateFormatter: 2016-01-28T17:30:00.000Z=>Thu Jan 28 18:30:00 GMT+01:00 2016
DefaultDateFormatter: 2016-01-08T14:35:39.000Z=>Fri Jan 08 15:35:39 GMT+01:00 2016
DefaultDateFormatter: 2016-02-27T14:30:00.000Z=>Sat Feb 27 15:30:00 GMT+01:00 2016
DefaultDateFormatter: 2016-02-03T16:00:00.000Z=>Wed Feb 03 17:00:00 GMT+01:00 2016
DefaultDateFormatter: 2016-02-27T16:00:00.000Z=>Sat Feb 27 17:00:00 GMT+01:00 2016
DefaultDateFormatter: 2016-02-03T17:00:00.000Z=>Fri Jan 01 01:00:00 GMT+01:00 2016 <=BAD!
DefaultDateFormatter: 2016-01-27T09:42:47.000Z=>Wed Jan 27 10:42:47 GMT+01:00 2016
DefaultDateFormatter: 2016-01-08T14:35:39.000Z=>Fri Jan 08 15:35:39 GMT+01:00 2016
DefaultDateFormatter: 2016-02-03T17:00:00.000Z=>Wed Feb 03 18:00:00 GMT+01:00 2016
DefaultDateFormatter: 2016-04-20T08:00:00.000Z=>Wed Feb 03 09:00:00 GMT+01:00 2016 <=BAD!
DefaultDateFormatter: 2016-02-03T18:00:00.000Z=>Wed Feb 03 19:00:00 GMT+01:00 2016
DefaultDateFormatter: 2016-01-27T14:11:08.000Z=>Wed Apr 27 16:11:08 GMT+02:00 2016
DefaultDateFormatter: 2016-04-20T10:00:00.000Z=>Wed Apr 27 16:11:08 GMT+02:00 2016 <=BAD!
DefaultDateFormatter: 2016-01-12T14:34:20.000Z=>Wed Jan 27 01:34:20 GMT+01:00 2016 <=BAD!
DefaultDateFormatter: 2016-01-27T17:30:00.000Z=>Wed Jan 27 18:30:00 GMT+01:00 2016
DefaultDateFormatter: 2016-01-28T18:30:00.000Z=>Thu Jan 28 19:30:00 GMT+01:00 2016
DefaultDateFormatter: 2016-02-27T13:30:00.000Z=>Sat Feb 27 14:30:00 GMT+01:00 2016
DefaultDateFormatter: 2016-01-25T17:49:36.000Z=>Mon Jan 25 18:49:36 GMT+01:00 2016
DefaultDateFormatter: 2016-02-27T15:00:00.000Z=>Sat Feb 27 16:00:00 GMT+01:00 2016
DefaultDateFormatter: 2016-01-25T17:30:00.000Z=>Mon Jan 25 18:30:00 GMT+01:00 2016
DefaultDateFormatter: 2016-01-08T14:35:39.000Z=>Fri Jan 08 15:35:39 GMT+01:00 2016
DefaultDateFormatter: 2016-01-25T18:30:00.000Z=>Mon Jan 25 19:30:00 GMT+01:00 2016
```
